### PR TITLE
[AI Rewriter]: Implement ability to rewrite text

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves https://github.com/brave/brave-browser/issues/38660

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/ai_rewriter/resources/page/components/Layout.tsx [hunk 1]
```diff
@@ -75,38 +65,41 @@ const Root = styled.div`
...
         <Icon name="arrow-redo" />
       </Button>
-      <Button fab kind="outline" isDisabled={!canErase} onClick={erase}>
+      <Button fab kind="outline" isDisabled={!context.canErase} onClick={context.erase}>
         <Icon name="erase" />
       </Button>
       <FlexSpacer />
-      <Button onClick={generate}>
-        Generate
+      <Button isLoading={context.isGenerating} onClick={context.generatedText ? context.acceptGeneratedText : context.submitRewriteRequest} isDisabled={context.isGenerating}>
```
Comment (sangwoo108) [inline components/ai_rewriter/resources/page/components/Layout.tsx:97]: I'm wondering if we should also disable the button when
* prompt is empty
* tool isn't selected.

File: components/ai_chat/resources/page/components/tools_button_menu/index.tsx [hunk 2]
```diff
@@ -9,27 +9,30 @@ import classnames from '$web-common/classnames'
...
+          props.isToolsMenuOpen && props.inputText.startsWith('/')
       })}
-      isOpen={context.isToolsMenuOpen}
-      onClose={() => context.setIsToolsMenuOpen(false)}
+      isOpen={props.isToolsMenuOpen}
+      onClose={() => props.setIsToolsMenuOpen(false)}
     >
       <div slot='anchor-content'>{props.children}</div>
-      {context.actionList.map((actionGroup) => {
+      {props.actionList.map((actionGroup) => {
```
Comment (sangwoo108) [inline components/ai_chat/resources/page/components/tools_button_menu/index.tsx:35]: It seems react complains about key here

```
[22467:259:0530/151355.271979:ERROR:CONSOLE(35307)] "Warning: Each child in a list should have a unique "key" prop.%s%s See https://reactjs.org/link/warning-keys for more information.%s 
Check the render method of `ToolsButtonMenu`.  
    at div
    at ToolsButtonMenu (chrome://rewriter/rewriter_ui.bundle.js:59754:83)
```

File: components/ai_rewriter/resources/page/components/Layout.tsx [hunk 3]
```diff
@@ -75,38 +65,41 @@ const Root = styled.div`
...
         </Button>
       </DialogControls>
     </TitleRow>
-    <FiltersContainer>
-      <InitialText />
-    </FiltersContainer>
-    <NoContent />
+    {context.generatedText
+      ? <ModifyGeneration />
+      : <BeginGeneration />}
```
Comment (sangwoo108) [inline components/ai_rewriter/resources/page/components/Layout.tsx:85]: Just out of curiosity, context.generatedText is working now? On my local env I wasn't able to see <ModifyGeneration> but undo button was enabled, which feels like they out of sync.

**Review by sangwoo108 — APPROVED** [review]
LGTM

File: components/ai_rewriter/resources/page/components/Layout.tsx (hunk 3 above)
Comment (fallaciousreasoning) [reply components/ai_rewriter/resources/page/components/Layout.tsx:89]: I think that might mean there was an error that came back - I updated this to not push to history unless the new generated text is different to the old.

**Comment by fallaciousreasoning** [comment]
**Note:** Current master is outputting the `</response>` tag (this is happening without this branch, and via the old context menu system)

**Review by mkarolin — APPROVED** [review]
`chromium_src` and `DEPS` lgtm

**Comment by petemill** [comment]
@fallaciousreasoning I think it would be useful to add some screenshots to this or the issue, as well as start a more thorough test plan.

**Comment by fallaciousreasoning** [comment]
**Note:** This only works when the new conversation api is disabled `--enable-features=AIChat:conversation_api\/false`

**Pull request merged**